File: 4.8-bc/go/stage-3-hexagonal/notifications/app/service.go
package app

import (
	"context"
	"time"

	"github.com/google/uuid"

	"github.com/genkovich/claude-course-demos/4.8-bc/go/stage-3-hexagonal/notifications/domain"
)

type Service struct {
	repo   domain.Repository
	sender domain.Sender
}

func New(repo domain.Repository, sender domain.Sender) *Service {
	return &Service{repo: repo, sender: sender}
}

// Send — головний use case. Створює notification, шле через Sender, зберігає у БД.
// Викликається і HTTP handler-ом, і event subscriber-ами.
func (s *Service) Send(ctx context.Context, userID uuid.UUID, channel domain.Channel, payload string) (*domain.Notification, error) {
	n := domain.Notification{
		ID:      uuid.New(),
		UserID:  userID,
		Channel: channel,
		Payload: payload,
	}
	if err := s.sender.Send(ctx, n); err != nil {
		return nil, err
	}
	now := time.Now().UTC()
	n.SentAt = &now
	if err := s.repo.Save(ctx, n); err != nil {
		return nil, err
	}
	return &n, nil
}

File: 4.8-bc/ts/stage-3-hexagonal/src/notifications/app/service.ts
// Notifications use case — головний Send. Викликається і HTTP handler-ом,
// і event subscriber-ами.
import { randomUUID } from "node:crypto";

import type {
  Channel,
  Notification,
  Sender,
} from "../domain/notification.js";
import type { Repository } from "../domain/repository.js";

export class Service {
  constructor(
    private readonly repo: Repository,
    private readonly sender: Sender,
  ) {}

  async send(
    userId: string,
    channel: Channel,
    payload: string,
  ): Promise<Notification> {
    const n: Notification = {
      id: randomUUID(),
      userId,
      channel,
      payload,
      sentAt: null,
    };
    await this.sender.send(n);
    n.sentAt = new Date();
    await this.repo.save(n);
    return n;
  }
}
